### PR TITLE
[jjo] diff: handle empty config values

### LINKED ISSUE #179: diff should skip empty config values, as api won't return them
Example from creating a pod from scratch (note empty `env`):
```
$ cat nginx.pod.yaml 
apiVersion: v1
kind: Pod
metadata:
  labels:
    run: nginx
  name: nginx
spec:
  containers:
  - image: nginx
    env: []
    name: nginx
  restartPolicy: Never

$ kubecfg update -v nginx.pod.yaml
INFO  Updating pods nginx  
INFO  Creating non-existent pods nginx

$ kubecfg diff --diff-strategy subset nginx.pod.yaml
---
- live pods nginx
+ config pods nginx
 {
   "apiVersion": "v1",
   "kind": "Pod",
   "metadata": {
     "labels": {
       "run": "nginx"
     },
     "name": "nginx"
   },
   "spec": {
     "containers": [
       {
         "image": "nginx",
         "imagePullPolicy": "IfNotPresent",
         "name": "nginx",  
         "resources": {
         }
+        "env": [
+        ]
       }
     ],
     "dnsPolicy": "ClusterFirst",
     "restartPolicy": "Never"
   }
 }
ERROR Differences found.  
```

### PR DESCRIPTION
**Pull request by jjo**
Fixes #179.

### CODE REVIEW

File: pkg/kubecfg/diff.go [hunk 1]
```diff
@@ -102,6 +103,26 @@ func (c DiffCmd) Run(apiObjects []*unstructured.Unstructured, out io.Writer) err
 	return nil
 }
 
+// isEmptyValue is taken from the encoding/json package in the
```
Comment (mmikulicic) [inline pkg/kubecfg/diff.go]: please mention https://github.com/golang/go/issues/7501

File: pkg/kubecfg/diff.go [hunk 2]
```diff
@@ -118,6 +139,9 @@ func removeMapFields(config, live map[string]interface{}) map[string]interface{}
 	for k, v1 := range config {
 		v2, ok := live[k]
 		if !ok {
+			if isEmptyValue(reflect.ValueOf(v1)) {
```
Comment (mmikulicic) [inline pkg/kubecfg/diff.go]: perhaps a comment wouldn't hurt here (so one doesn't have to git blame through the full future chain of commits until one hits this PR with the relevant context)

File: pkg/kubecfg/diff.go (hunk 1 above)
Comment (jjo) [reply pkg/kubecfg/diff.go]: +1, done.

File: pkg/kubecfg/diff.go (hunk 2 above)
Comment (jjo) [reply pkg/kubecfg/diff.go]: +1, done.

File: pkg/kubecfg/diff.go [hunk 3]
```diff
@@ -102,6 +103,28 @@ func (c DiffCmd) Run(apiObjects []*unstructured.Unstructured, out io.Writer) err
 	return nil
 }
 
+// isEmptyValue is taken from the encoding/json package in the
+// standard library.
+// See also feature request for golang reflect pkg at
+// https://github.com/golang/go/issues/7501
+func isEmptyValue(v reflect.Value) bool {
+	switch v.Kind() {
```
Comment (anguslees) [inline pkg/kubecfg/diff.go]: Fwiw, this only has to deal with the types that `encoding/json` produces, so a simple typeswitch will be faster:

```golang
func isEmptyValue(i interface{}) bool {
        switch v := i.(type) {
        case []interface{}:
                return len(v) == 0
        case map[string]interface{}:
                return len(v) == 0
        case bool:
                return !v
        case float64:
                return v == 0
        case string:
                return v == ""
        case nil:
                return true
        default:
                panic("Found unexpected type %T in json unmarshal (value=%v)", i, i)
        } 
}
```

Comment (jjo) [reply pkg/kubecfg/diff.go]: @anguslees: nopes, that won't support nested `nil`s - I've augmented
`diff_test.go` a bit to exercise that also.

FWIW using above simpler implementation would expectedly fail with:
```
# github.com/ksonnet/kubecfg/pkg/kubecfg
pkg/kubecfg/diff.go:142:8: too many arguments to panic: panic("Found unexpected type %T in json unmarshal (value=%v)", i, i)
Makefile:35: recipe for target 'kubecfg' failed
```

Comment (jjo) [reply pkg/kubecfg/diff.go]: @anguslees nvm on above reply (early morning low caffeine disclaimer :P),
re-implemented it using your suggestions above (+couple nits).

**Pull request merged**